pkg/appsec/appsec.go

```diff
@@ -0,0 +1,133 @@
+			"`appsec_config.enabled: true` to your datadog.yaml file")
+		return disabled("appsec agent disabled by configuration"), nil
+	}
+	return newIntakeReverseProxy(cfg.IntakeURL, cfg.APIKey, cfg.MaxPayloadSize, transport), nil
+}
+
+// newIntakeReverseProxy creates a reverse proxy to the intake backend using the
+// given transport round-tripper.
+// The reverse proxy handler also limits the request body size and adds extra
+// headers such as Dd-Api-Key and Via.
+func newIntakeReverseProxy(target *url.URL, apiKey string, maxPayloadSize int64, transport http.RoundTripper) http.Handler {
+	proxy := httputil.NewSingleHostReverseProxy(target)
+	via := fmt.Sprintf("trace-agent %s", info.Version)
+	// Wrap and overwrite the returned director to add extra headers
+	director := proxy.Director
+	proxy.Director = func(req *http.Request) {
+		// Call the original director changing the request target
+		director(req)
+		// Set extra headers
+		req.Header.Set("Via", via)
+		req.Header.Set("Dd-Api-Key", apiKey)
+	}
+	proxy.ErrorHandler = func(w http.ResponseWriter, req *http.Request, err error) {
+		w.Header().Add("Content-Type", "application/json")
+		w.WriteHeader(http.StatusInternalServerError)
+		if err := json.NewEncoder(w).Encode(err); err != nil {
+			log.Error(err)
+		}
+	}
+	proxy.Transport = withMetrics(transport, maxPayloadSize)
+	proxy.ErrorLog = stdlog.New(logutil.NewThrottled(5, 10*time.Second), "Appsec backend proxy: ", 0)
+	return proxy
+}
+
+const (
+	appSecRequestMetricsPrefix        = "datadog.trace_agent.appsec."
+	appSecRequestCountMetricsID       = appSecRequestMetricsPrefix + "request"
+	appSecRequestDurationMetricsID    = appSecRequestMetricsPrefix + "request_duration_ms"
+	appSecRequestPayloadSizeMetricsID = appSecRequestMetricsPrefix + "request_payload_size"
+	appSecRequestErrorMetricsID       = appSecRequestMetricsPrefix + "request_error"
+)
+
+// metricsTags returns the metrics tags of a request.
+func metricsTags(req *http.Request) []string {
+	tags := []string{"path:" + req.URL.Path}
+	if ct := req.Header.Get("Content-Type"); ct != "" {
+		tags = append(tags, "content_type:"+ct)
+	}
+	return tags
+}
+
+type roundTripper struct {
+	http.RoundTripper
+	maxPayloadSize int64
+}
+
+func withMetrics(rt http.RoundTripper, maxPayloadSize int64) http.RoundTripper {
+	return &roundTripper{
+		RoundTripper:   rt,
+		maxPayloadSize: maxPayloadSize,
+	}
+}
+
+// RoundTrip limits the request body size that can be read and performs internal monitoring metrics
+func (r *roundTripper) RoundTrip(req *http.Request) (res *http.Response, err error) {
+	var lr *apiutil.LimitedReader
+	if req.Body != nil && r.maxPayloadSize > 0 {
+		lr = apiutil.NewLimitedReader(req.Body, r.maxPayloadSize)
+		req.Body = lr
+	}
+
+	now := time.Now()
+	defer func() {
+		tags := metricsTags(req)
+		if lr != nil {
+			metrics.Histogram(appSecRequestPayloadSizeMetricsID, float64(lr.Count), tags, 1)
```

Comment:
This scares me a little. A race condition (panic) is possible here. See: https://github.com/golang/go/blob/go1.16/src/net/http/client.go#L134-L138.

We have 2 options:
* Alter the `LimitedReader` struct to access `Count` under a lock (and then retrieve it using a method `Count()`)
* Give up on this metric entirely

Reply:
Removed as we couldn't find a proper way of making LimitedReader atomic.